File: src/youtube-api/services/youtube-subscription-api.service.ts
import { injectable, inject } from "inversify";
import { Types } from "../../types";

import { GaxiosResponse } from "gaxios";

import { OAuth2Client } from "googleapis-common";

import { Youtube, Schema$Subscription } from "../types/youtube.aliases";

import { YOUTUBE_RESOURCE_KIND } from "../metadata/youtube-resource-kind.enum";

@injectable()
export class YoutubeSubscriptionApiService {
  constructor(
    @inject(Types.OAuth2Client)
    private client: OAuth2Client,
    @inject(Types.Youtube)
    private youtube: Youtube
  ) {}

  public addSubscription(
    channelId: string
  ): Promise<GaxiosResponse<Schema$Subscription>> {
    return this.youtube.subscriptions.insert({
      auth: this.client,
      part: "snippet",
      requestBody: {
        snippet: {
          resourceId: {
            kind: YOUTUBE_RESOURCE_KIND.CHANNEL,
            channelId: channelId
          }
        }
      }
    });
  }

  public hasSubscription(channelId: string): Promise<boolean> {
    return new Promise((resolve, reject) =>
      this.youtube.subscriptions
        .list({
          auth: this.client,
          mine: true,
          forChannelId: channelId
        })
        .then(response => {
          if (response.data.items) {
            return resolve(response.data.items.length > 0);
          }
          resolve(false);
        })
        .catch(error => reject(error))
    );
  }
}

File: packages/youtube-api/lib/src/services/youtube_subscription_api.service.dart
import 'package:googleapis/youtube/v3.dart';
import 'package:rxdart/subjects.dart';

import 'package:youtube_api/src/metadata/youtube_resource_kind.dart';

class YoutubeSubscriptionApiService {
  final YoutubeApi _youtube;

  YoutubeSubscriptionApiService(this._youtube);

  Subject<Subscription> addSubscriptions(List<String> channelIds) {
    final subject = ReplaySubject<Subscription>();

    Future.forEach(
            channelIds,
            (channelId) => addSubscription(channelId)
                .then((subscription) => subject.add(subscription))
                .catchError((error) => subject.addError(error)))
        .then((onComplete) => subject.close());

    return subject;
  }

  Future<Subscription> addSubscription(String channelId) {
    final request = Subscription.fromJson({
      "snippet": SubscriptionSnippet.fromJson({
        "resourceId": ResourceId.fromJson({
          "kind": YOUTUBE_RESOURCE_KIND_PROPS[YOUTUBE_RESOURCE_KIND.CHANNEL],
          "channelId": channelId
        })
      })
    });

    return _youtube.subscriptions.insert(request, "snippet");
  }

  Future<bool> hasSubscription(String channelId) {
    return _youtube.subscriptions
        .list("", mine: true, forChannelId: channelId)
        .then((response) {
      return response.items.isNotEmpty;
    });
  }
}
